zng/union.go

```diff
@@ -0,0 +1,87 @@
+package zng
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/mccanne/zq/zcode"
+)
+
+type TypeUnion struct {
+	id    int
+	Types []Type
+}
+
+func NewTypeUnion(id int, types []Type) *TypeUnion {
+	return &TypeUnion{id, types}
+}
+
+func (t *TypeUnion) ID() int {
+	return t.id
+}
+
+func (t *TypeUnion) SetID(id int) {
+	t.id = id
+}
+
+func (t *TypeUnion) TypeIndex(index int) (Type, error) {
+	if index < 0 || index+1 > len(t.Types) {
+		return nil, ErrUnionIndex
+	}
+	return t.Types[index], nil
+
+}
+
+func (t *TypeUnion) String() string {
+	var ss []string
+	for _, typ := range t.Types {
+		ss = append(ss, typ.String())
+	}
+	return fmt.Sprintf("union[%s]", strings.Join(ss, ","))
+}
+
+func (t *TypeUnion) Parse(in []byte) (zcode.Bytes, error) {
+	c := bytes.Index(in, []byte{':'})
+	if c < 0 {
+		return nil, ErrBadValue
+	}
+	index, err := strconv.Atoi(string(in[0:c]))
+	if err != nil {
+		return nil, err
+	}
+	typ, err := t.TypeIndex(index)
+	if err != nil {
+		return nil, err
+	}
+	out := zcode.AppendUvarint(zcode.Bytes{}, uint64(index))
```

Comment:
Nit: s/zcode.Bytes{}/nil/ unless I'm missing something and that won't compile.

Reply:
done